src/axom/core/Array.hpp

```diff
@@ -55,11 +88,152 @@ bool operator==(const Array<T>& lhs, const Array<T>& rhs);
  * \return true if the Arrays do not have the same allocator ID, are not of
  * equal length, or do not have the same elements.
  */
-template <typename T>
-bool operator!=(const Array<T>& lhs, const Array<T>& rhs);
+template <typename T, IndexType DIM>
+bool operator!=(const Array<T, DIM>& lhs, const Array<T, DIM>& rhs);
 
 /// @}
 
+/*!
+ * \brief Policy class for implementing Array behavior that differs
+ * between the 1D and multidimensional cases
+ * 
+ * \tparam T The element/value type
+ * \tparam DIM The dimension of the Array
+ */
+template <typename T, IndexType DIM>
+class ArrayImpl
+{
+public:
+  using ArrayType = Array<T, DIM>;
+
+  /*!
+   * \brief Parameterized constructor that sets up the default strides
+   *
+   * \param [in] args the parameter pack of sizes in each dimension.
+   */
+  template <typename... Args>
+  ArrayImpl(Args... args) : m_dims {static_cast<IndexType>(args)...}
+  {
+    // Row-major
+    m_strides[DIM - 1] = 1;
+    for(int i = static_cast<int>(DIM) - 2; i >= 0; i--)
+    {
+      m_strides[i] = m_strides[i + 1] * m_dims[i + 1];
+    }
+  }
+
+  /*!
+   * \brief Dimension-aware accessor, returns a reference to the given value.
+   *
+   * \param [in] args the parameter pack of indices in each dimension.
+   *
+   * \note equivalent to *(array.data() + idx).
+   *
+   * \pre sizeof...(Args) == DIM
+   * \pre 0 <= args[i] < m_dims[i] for i in [0, DIM)
+   */
+  template <typename... Args,
+            typename SFINAE = typename std::enable_if<sizeof...(Args) == DIM>::type>
+  T& operator()(Args... args)
+  {
+    IndexType indices[] = {static_cast<IndexType>(args)...};
+    IndexType idx =
+      std::inner_product(indices, indices + DIM, m_strides.begin(), 0);
```

Comment:
``std::inner_product``

Future/GPU concern: This might need to be rewritten when porting to GPUs

Reply:
Logged in #633